### ISSUE TITLE
Documentation thread - Any help is most welcome!

### ISSUE THREAD

**Issue by jaffasplaffa** [plugdata-team/plugdata#75]
Hey :) 

This issue is dedicated to documenting PlugData and we hope that community is willing to help out!

Documentation includes short descriptions of object functionality, as well as in/output descriptions. The idea is that these description will show when you hover the mouse over objects and in/outputs.

A better and more specific description of the tasks that help is needed for, will be presented soon........

If you feel like helping out, please drop a comment. No coding experience is needed, but the ability to explain something briefly and easy understandable is a great skill to master for such a task :) 

Thanks!

**Comment by timothyschoen** [comment]
There are currently about 200 pddoc files that need documentation, and 3 people that are willing to do it so far (@jaffasplaffa @ludnny and I). How about we start by doing 50 each?

abs        (A)  - drawtext (D): @jaffasplaffa 
element (E)  - mult~.     (M): @timothyschoen 
neq        (N)  - rzero~.    (R): @ludnny

If anyone else wants to contribute of anyone has more time available, we can start covering the last 50 objects as well.

No pressure btw, take all the time you need or let me know if 50 is too much. I just want to prevent that everyone will start working alphabetically from A ;)

I suggest you submit a pull request, then you'll appear on the page as a contributor.

For anyone who hasn't read about this yet: 

https://github.com/timothyschoen/PlugData/tree/main/Resources/pddoc/vanilla

Contains a bunch of xml files that need to have their "inlets", "outlets", and "description" fields filled in, I've attached an example of some finished ones below:

[examples.zip](https://github.com/timothyschoen/PlugData/files/8565199/examples.zip)

**Comment by timothyschoen** [comment]
Useful resources:

https://puredata.info/docs/tutorials/pd-refcard


<img width="194" alt="Screenshot 2022-04-26 at 18 47 04" src="https://user-images.githubusercontent.com/44585538/165351432-7cc0dd91-0ef4-47bf-a794-d33af76da65f.png">

In many help files leads to:

<img width="569" alt="Screenshot 2022-04-26 at 18 47 30" src="https://user-images.githubusercontent.com/44585538/165351482-13f98d9b-eaba-4d71-be9b-5f54c576cfe7.png">

**Comment by jaffasplaffa** [comment]
I am up for giving it a go :) 

But I am really not an expert on this. I need to be shown, preferable like I was an 8 year old kid, where to put the descriptions. 

I think I need to see a couple of already done descriptions, maybe with the descriptions highlighted or in fat text?

Just to make sure I don't waste a lot of out time on doing something the wrong way.

**Comment by timothyschoen** [comment]
Good idea, I'll provide some very clear instructions soon!

**Comment by jaffasplaffa** [comment]
Here is what I THINK I need to do. This is the "change" object:

<?xml version="1.0" encoding="utf-8"?>
<pddoc version="1.0">
    <object name="change">
        <title>change</title>
        <meta>
            <authors>
                <author>Miller Smith Puckette</author>
            </authors>
            <description>remove repeated numbers from a stream</description>
            <license>BSD</license>
            <library>puredata</library>
            <category>general</category>
            <keywords>general change</keywords>
            <since>0.27</since>
        </meta>
        <info>
            <par>-</par>
        </info>
        <arguments>
            <argument name="ARG_NAME" type="symbol">-</argument>
        </arguments>
        <methods>
            <method name="NAME">
            <param type="float" name="V" required="true">-</param></method>
        </methods>
        <inlets>
            <inlet>
                <xinfo on="bang">**output last value**</xinfo>
                <xinfo on="float"> **float input** </xinfo>
                <xinfo on="symbol">**symbol input**</xinfo>
            </inlet>
        </inlets>
        <outlets>
            <outlet> **output only if input changes** </outlet>
        </outlets>
        <example>
            <pdascii>
<![CDATA[

[change]

]]>
            </pdascii>
        </example>
    </object>
</pddoc>




I've highlighted the changes I've made.

**Comment by timothyschoen** [comment]
Yeah looks good! I'd make some things a bit more brief, like:

"output only if input changes" -> "output when input changes"

There's no arguments or methods yet. Maybe it would be nice to cover arguments as well, they could show up in the suggestions box when you've completed typing the name of the object?

Like this:
```<argument name="Initial Value" type="float" default="0"></argument>```

I found this info in the pd reference:

<img width="558" alt="Screenshot 2022-04-26 at 19 24 26" src="https://user-images.githubusercontent.com/44585538/165357547-b8de33b8-8ccc-446a-8272-b73fac9cdc93.png">

If there's no default in the pd reference, it can be omitted for those objects.

**Comment by timothyschoen** [comment]
I'd go with 

 ``` <xinfo on="set [float]">Set the value</xinfo> ```

The pd reference says  "set <float>" but that's very confusing in xml documents.

Another exception is objects like sel which have an uncertain number of inlets or outlets. So lets define that idea for our format:

```
<outlets>
  <xinfo repeating="1" on="Bang if input machtes argument $1">-</xinfo>
</outlets>
```

The repeating flag enables this behaviour. The $1 will be replaced with the inlet index.

**Comment by jaffasplaffa** [comment]
I did a Fork of the rep and did the FIRST object, abs:
https://github.com/jaffasplaffa/PlugData/blob/main/Resources/pddoc/vanilla/abs.pddoc

Removed arguments, since there are no arguments for that object.

(Sorry for all the questions, I just need to be confident I started properly :) ).

**Comment by timothyschoen** [comment]
Very nice, and no problem, it's better to make sure you get it right than to do a lot of work only to find out it's useless ;)

Small point is that's I'd make the object description slightly shorter if possible. Just "Absolute value" could maybe do in the case?

<img width="558" alt="Screenshot 2022-04-26 at 20 26 49" src="https://user-images.githubusercontent.com/44585538/165367434-acaae176-5006-4fdc-8190-f603fec6de2a.png">

I'd like to show them in the suggestions as well, so if it fits, that'd be nice.  It's not super important, but still a nice. I'll make the default suggestionbox size bigger as well.

**Comment by jaffasplaffa** [comment]
Yes I understand, I'll make it shorter. I just used the description from abs~. 

Personally I like the descriptions a bit more explanatory, cause I am not a coder, so I can actually understand what the object does, without having to do external research on Google. 

But yeah if its needed elsewhere and need short format, that's what we are gonna do :)

**Comment by timothyschoen** [comment]
@ludnny I accepted your pull request, but you're not appearing as a contributor yet (which I think would be well deserved), no idea why this is though. Maybe it takes some time to update...

**Comment by ludnny** [comment]
> ![Screen Recording 2022-04-27 at 11 11 42](https://user-images.githubusercontent.com/44585538/165484758-5232c94a-b6d6-4058-bb1f-122061aaa4ed.gif)

Waaw! This feature will change everything :)

**Comment by timothyschoen** [comment]
Yeah I agree, it helps a lot! Especially for objects with a many inlets/outlets it's a huge help. I also think it has some educational value for people who are new to pd. Once we cover ELSE and cyclone as well, it can also help people who aren't familiar with those libraries yet. Maybe @porres could assist with that?

Also, I'm wondering if we need to add a type to the output label? For metro we could have

``` <outlet  out="bang"> Metronome ticks </outlet> ```

So it'll read "(bang) Metronome ticks". I think that'll be even better. For objects that don't consistently output one type, it can be omitted.

I'm extending the pddoc format with new ideas here a bit, but adding extra attributes shouldn't break compatibility. I'll probably submit everything to the [pddoc project](https://github.com/uliss/pddoc) when it's done.

**Comment by ludnny** [comment]
>  I also think it has some educational value for people who are new to pd. 

That could break a huge barrier for begginners, I stopped using PD multiple times because of this barrier. 

> Also, I'm wondering if we need to add a type to the output label? For metro we could have
> 
> `<outlet out="bang"> Metronome ticks </outlet>`

For signals the type is signal?
`<outlet out="signal">Signal</outlet>`

**Comment by timothyschoen** [comment]
Yeah exactly!

**Comment by timothyschoen** [comment]
Sometimes I run into an object that doesn't exist in pd (anymore?), in that case I just remove it.

**Comment by ludnny** [comment]
> Sometimes I run into an object that doesn't exist in pd (anymore?), in that case I just remove it.

Ok, I remove the ones I can't find in Vanilla. (**_neq_**...)

For the object **_or_**, I assume it's the object **||** in PureData?
Should we remame that ||.pddoc?

**Comment by ludnny** [comment]
Do you want to encode the flags as well, in addition to the arguments?
(for example `netsend -u`)

What would be the format for that?

**Comment by timothyschoen** [comment]
good questions, I think "or" is actually "bitwise or" instead of "logical or", since the original description is "bit twiddling". There might still be some missing objects, I don't see a description for logical or, and GUI objects also appear missing.

About flags, I'd add ```flag="u"``` so that makes:

``` <argument name="ARG_NAME" type="symbol" flag="u">-</argument> ```

I still have to implement args in the UI, but I'll make sure it supports this!

**Comment by ludnny** [comment]
> good questions, I think "or" is actually "bitwise or" instead of "logical or", since the original description is "bit twiddling". There might still be some missing objects, I don't see a description for logical or, and GUI objects also appear missing.
> 
But if I want to do a pddoc for the object ||, how I should call it? As these characters are not allowed for filenames?
You link them manually, or based on their names?

**Comment by timothyschoen** [comment]
I've decided to make the syntax for using inlet index numbers a bit better, I hope it doesn't affect any work you've already done...

$mth  ->  index of inlet/outlet starting from 0
$nth  ->  index of inlet/outlet starting from 1
$arg  ->  object argument with the same index as the inlet/outlet

You can just insert these into any description, and it will be replaced with the correct value.

For example, [expr] looks like this:

```

        <inlets>
            <inlet repeating="1">
                <xinfo on="float">Replaces $f$nth in expression</xinfo>
                <xinfo on="symbol">Replaces $s$nth in expression</xinfo>
            </inlet>
        </inlets>

```

which results in:

<img width="487" alt="Screenshot 2022-04-27 at 14 40 50" src="https://user-images.githubusercontent.com/44585538/165520091-e8738390-fc27-483d-8c9d-3e7670aca65f.png">

**Comment by timothyschoen** [comment]
> > good questions, I think "or" is actually "bitwise or" instead of "logical or", since the original description is "bit twiddling". There might still be some missing objects, I don't see a description for logical or, and GUI objects also appear missing.
> 
> But if I want to do a pddoc for the object ||, how I should call it? As these characters are not allowed for filenames? You link them manually, or based on their names?

The name inside the <object name="||"> will be used, not the filename, so it doesn't really matter. I think calling one file bitor and the other logicor is the most clean solution.

**Comment by timothyschoen** [comment]
I think flags are still a bit confusing, but if we just make sure they all have an attribute flag at least it will be easy to fix them later.

**Comment by ludnny** [comment]
For inlets that could receive any type, I used the type `anything` (this is the type `a`, used in the `pack` object). 


```
            <inlet>
                <xinfo on="anything">Anything to send to parent patch</xinfo>
            </inlet>
```
I'm not sure it's an official type in pd?

Or I could do this:

```
            <inlet>
                <xinfo on="bang, float, symbol, list">Anything to send to parent patch</xinfo>
            </inlet>
```

**Comment by timothyschoen** [comment]
I think anything is fine, for documentation purposes it's at least much cleaner.

**Comment by ludnny** [comment]
I think there is a problem for **outlets** in this format. 
How do you see the difference between 3 outlets, and 2 outlets with one having 2 different types? :
```
        <outlets>
            <outlet out="symbol">Messages sent from connected netsend objects</outlet>
            <outlet out="float">Number of open connections for TCP connections</outlet>
            <outlet out="symbol">Address and port</outlet>
        </outlets>
```

It would be more logical to use the same structure as  **inlets**, don't you think?

(here is a fake example to show a situation with 2 outlets, and 3 types) :

```
        <outlets>
            <outlet>
                <xinfo on="symbol">Messages sent from connected netsend objects</outlet>
                <xinfo on="float">Number of open connections for TCP connections</outlet>
            </outlet>
            <outlet>
                <xinfo on="list">Address and port</outlet>
            </outlet>
        </outlets>
```

**Comment by ludnny** [comment]
Oups, I made a mistake, should close with `/xinfo`:

    <outlets>
            <outlet>
                <xinfo on="symbol">Messages sent from connected netsend objects</xinfo>
                <xinfo on="float">Number of open connections for TCP connections</xinfo>
            </outlet>
            <outlet>
                <xinfo on="list">Address and port</xinfo>
            </outlet>
        </outlets>

**Comment by ludnny** [comment]
and `out` rather than `on`:

    <outlets>
        <outlet>
            <xinfo out="symbol">Messages sent from connected netsend objects</xinfo>
            <xinfo out="float">Number of open connections for TCP connections</xinfo>
        </outlet>
        <outlet>
            <xinfo out="list">Address and port</xinfo>
        </outlet>
    </outlets>

**Comment by ludnny** [comment]
>  I'm gonna have to redo some objects ;)

Don't lose your time on that, we will manage the Vanilla objects with @jaffasplaffa, that should be fast once we have the good format!

For the other objects (ELSE, Cyclone...), I think that would be safer to wait for these pddocs to be tested and re-tested in multiple contexts.

**Comment by jaffasplaffa** [comment]
Hey :) 

Wow, so much to catch up on. I did make the 2 first objects yesterday, abs & abs~ but then I started running into questions.

adc~ was the first one, it has amount of input that can be set by argument. I need to read all this again, I think but it quickly got a bit overwhelming.

Then I tried and.pddoc too. There are two and objects in Pd & and &&. I think there need to be individual explanations for those two, no?

**Comment by timothyschoen** [comment]
Let's definitely make sure it works first! About multiple contexts: I wonder how CEAMMC will feel about the syntax changes we've made, they probably won't accept the pull request like this because it won't work with their current scripts. We could still pitch this to pd-vanilla, we'll see what they think about it.

Cyclone is manageable, it has around 190 objects. ELSE is a different story, it has like 450 objects. I'm also very busy with solving other things, but I'd feel bad for letting you handle all this work, because it's not very fun to do (even though the result will be great).

**Comment by ludnny** [comment]
> Hey :)
> 
> Wow, so much to catch up on.

Yes! More and more specific situations that change the rules... 

> adc~ was the first one, it has amount of input that can be set by argument. I need to read all this again, I think but it quickly got a bit overwhelming.

Don't hesitate to let the difficult ones at the moment (the ones with dynamic numbers of input/output), to focus on the easiest. I do that, to not be stopped in the process, 

> 
> Then I tried and.pddoc too. There are two and objects in Pd & and &&. I think there need to be individual explanations for those two, no?

`and.pddoc` is for `&&` 
`bitwiseand.pddoc` is for `&`

**Comment by timothyschoen** [comment]
> Hey :)
> 
> Wow, so much to catch up on. I did make the 2 first objects yesterday, abs & abs~ but then I started running into questions.
> 
> adc~ was the first one, it has amount of input that can be set by argument. I need to read all this again, I think but it quickly got a bit overwhelming.
> 
> Then I tried and.pddoc too. There are two and objects in Pd & and &&. I think there need to be individual explanations for those two, no?

Yes, sorry for all these complications, I honestly didn't see them coming! The &/&& situation is basically the same as the bit/logical or. I'd call the files "bitand.pddoc" and "logicand.pddoc". Inside the file, set the object name to && (logicand) or & (bitand).

For adc~, you have to use the repeating syntax, like this:

<outlets>
            <outlet repeating="1">
                <xinfo out="signal">Audio in channel: $arg </xinfo>
            </outlet>
 </outlets >


$arg will be replaced with the args that you pass along, so for [adc~ 2 4], $arg for the first inlet is 2, for the second inlet 4.
There's also $mth (inlet index from 0) and $nth (inlet index from 1)

But like @ludnny, skip them if you can't figure them out, it's no problem. Maybe some situations will require another new flag to be added, which makes it easier for me to handle those.

**Comment by ludnny** [comment]
> $arg will be replaced with the args that you pass along, so for [adc~ 2 4], $arg for the first inlet is 2, for the second inlet 4. There's also $mth (inlet index from 0) and $nth (inlet index from 1)

Could you give us an example of when using `$mth` or `$nth` ?
For example, what would you use for pack? `$nth` ?

**Comment by timothyschoen** [comment]
I'd almost always use $nth, because most humans count from 1. For pack you could even use $arg, if you want to show the symbol for what you're packing. I added $mth for completeness but I think it might not be necessary after all.

**Comment by timothyschoen** [comment]
> 
> `and.pddoc` is for `&&` `bitwiseand.pddoc` is for `&`

This is different from what I said but I like this as well, so if you consistently do this it's fine!

**Comment by porres** [comment]
are you including ceammc?

**Comment by timothyschoen** [comment]
I'm using ceammc's format for pd documentation, since it already has a basic file for each object with a description. But we had to change some things to be able to document all objects properly, so it's not really compatible anymore. I think that might be why there are some extra objects in there that are not in pd-vanilla.

**Comment by porres** [comment]
we're doing something similar for vanilla, hold on

**Comment by timothyschoen** [comment]
I'm very curious about this, I couldn't find anything about it

**Comment by porres** [comment]
we're dealing with it here https://github.com/pure-data/pddp

**Comment by porres** [comment]
https://github.com/pure-data/pddp/issues/46

**Comment by timothyschoen** [comment]
This looks great, and a large part is already documented! @jaffasplaffa and @ludnny: I'm sorry if this has caused you to do unnecessary work... I didn't know about this project. This format honestly looks much better than the pddoc format, and should be quicker to fill out for ELSE and cyclone as well.

This folder: https://git.iem.at/pd/pd.iem.sh/-/tree/main/content/objects

Contains a lot of what we need. But the good news is, we don't have to do work on vanilla objects, and we have an easier format for documenting ELSE and cyclone in the future. Thanks @porres!

**Comment by porres** [comment]
help us there :)

**Comment by timothyschoen** [comment]
Is documentation of externals also part of your goal? Because then I'd say we'd better contribute to that instead, I'll make small modifications for PlugData's tooltips, for things like brevity or showing actual indices and argument names.

**Comment by porres** [comment]
it'd be nice to have online references for my externals, but it would require too damn work :)

**Comment by timothyschoen** [comment]
I agree, and since your ELSE helpfiles are very neat, I think it can be done. If everyone is still in, we could slowly work our way through it.

I think that converting this:

<img width="548" alt="Screenshot 2022-04-29 at 02 53 18" src="https://user-images.githubusercontent.com/44585538/165869995-52d54e4d-290b-45d1-910d-a056ad1b6176.png">

to this:

<img width="396" alt="Screenshot 2022-04-29 at 02 54 04" src="https://user-images.githubusercontent.com/44585538/165870067-b9fb11a8-47aa-44ab-a95d-047175b4bd39.png">

is manageable (different object, but just for reference). Especially if you get a nice workflow where you use the documentation browser in vanilla to go through the helpfiles quickly. To get started, I'll export all the files we're gonna need programatically and I'll fill in the name already. Then it's just copy/pasting some descriptions. If we're all in to help, we can get it done!

**Comment by timothyschoen** [comment]
Autogenerated files are now here:

https://github.com/timothyschoen/PlugData/tree/main/Resources/pddp

**Comment by ludnny** [comment]
Nice format, should be easier to fill than the pddocs!

What does "draft: false" mean in the md?


So, what is the strategy?
@timothyschoen , you assign us some files?

Are the vanilla files all ok?
Maybe we wait for you to implement this system in PlugData, with the vanilla md files. 
So you can see what works and not (especially with particular cases, like dynamic inlet...). 
And after a first release with all vanilla object tested, we do the same for Cyclone and ELSE. 

About the actual status of ELSE (1.0 RC2):
> This project is still in an experimental phase (currently at a 'release candidate' phase), where changes may occur and backwards compatibility is not guaranteed until a final and more stable release is available. 
https://github.com/porres/pd-else

It looks like @porres is getting closer and closer to a stable 1.0 release, so we should wait for that before documenting 400 objects that are subject to change on a daily basis!

**Comment by timothyschoen** [comment]
Yeah I agree, not having to worry about brackets and stuff makes it a much faster process. 

I assume the draft flag is used to mark that a piece of documentation is unfinished. This is actually useful for us: if you get stuck on an object, just set the draft flag to true and we can easily find them later!

I'm already successfully parsing this format in PlugData as well. It's less clean than pddocs (which is easy because it's just xml), but at the same time I think parsing this manually might be more efficient because we don't need to use a complete xml parser.

I believe the idea of a release candidate is that the objects themselves won't change unless there's something seriously wrong. So we could probably get started pretty soon, @porres could probably tell us the details. We can already work on cyclone. I wanna try to get a whole bunch of them done tonight.

Here's what needs to happen for vanilla:

- Some object descriptions are not very complete, for example, all bitwise operators just say "bit twiddling" in their descriptions. And there's many things that could be worded better: metro's second inlet says "set metronome time for next tempo", which I think is not the clearest explanation.

I'll create a detailed post about the new format in a minute!

**Comment by timothyschoen** [comment]
So I've exported empty templates for all objects, that look like this:

<img width="396" alt="Screenshot 2022-04-29 at 13 28 36" src="https://user-images.githubusercontent.com/44585538/165936133-0eb6e028-0809-482d-b482-e1242faef90a.png">

I've added one change to the format: the default argument value (if applicable) is now more clearly specified.

You have to:

1. Fill the description
2. Possibly add categories
3. Fill arguments: each argument has a type, description and optionally a default. Normally, defaults are reported in the description
4. Fill inlet/outlet description
5. If the help file contains a long description, paste if at the bottom where it says "LONG DESCRIPTION HERE"


Special cases (I'll add to the if we find more):

1. dynamic inlets and outlets:

We need to change the format a little bit to get this to work. Normally, inlets/outlets are numbered with "1st:", "2nd:" etc. For dynamic inlets/outlets, use "nth:" instead. After that you can use $nth (replaced with inlet/outlet index) and  $arg (replaced with the corresponding argument of the object) inside the description.


Another big advantage is that this format can be converted into a reference website. For that situation, you can search for the word "nth" to change the dynamic descriptions to something more website-friendly. I think that explicitly adding default values for arguments will only improve the quality of the docs.

**Comment by porres** [comment]
I'm still in the process of revising ALL help files in Vanilla before we can copy them into html, check https://github.com/pure-data/pure-data/pull/1594 and my branch. https://github.com/porres/pure-data/tree/Documentation

You can help me revising the information, spot typos and mistakes in the documentation, that are plently but you need to pay lots of attention to find them. Once this revision is done, we can copy to html... the first copy was made just as a draft but I have made dozens of fixes after that!

**Comment by porres** [comment]
As for cyclone/ELSE, they already have this reference section, one could just copy as well, but I would really really like to have some help revising all of the help files searching for mistakes, examples that are not working or with bugs, typos, etc, etc... --> https://github.com/porres/pd-cyclone

Same for ELSE, which can come for last. --> https://github.com/porres/pd-else

**Comment by timothyschoen** [comment]
<img width="542" alt="Screenshot 2022-04-30 at 16 09 59" src="https://user-images.githubusercontent.com/44585538/166109006-48fc8a4c-6320-4219-900e-8433c4c54e4f.png">

Arguments will now be shown after typing an object name!

**Comment by porres** [comment]
how hard would it be to implement this as a plug in in vanilla?

**Comment by timothyschoen** [comment]
I think someone already did the object suggestions as a plugin right? I think it should be doable, but I have very little experience with tcl/tk so I wouldn't be able to implement it myself. If someone wanted to implement this, they could use PlugData as a reference for parsing the markdown files though.

**Comment by porres** [comment]
> I think someone already did the object suggestions as a plugin right? 

right, but not args

**Comment by timothyschoen** [comment]
I think I would be a pretty easy to extend it like that for someone who knows tcl/tk.

I was wondering, are you planning to release RC2 anytime soon? I'm not that far away from releasing v0.5.1, I'm wondering if it's worth waiting for.

**Comment by porres** [comment]
> are you planning to release RC2 anytime soon?

no

**Comment by porres** [comment]
I now want to include fluidsynth~ in ELSE, if you can help me figure it out it'd be great, and it might also be a challenge to include it in plugdata https://github.com/porres/pd-fluidsynth

see my emails to the pd dev list

**Comment by porres** [comment]
once we figure it out I can release an update of ELSE

**Comment by timothyschoen** [comment]
That'd be cool! I'm going to finish version 0.5.1 off first (I'm pretty close I think), after that I might have some time to look at fluidsynth~.

Does it have some sort of GUI? If not, it should be very easy to include in PlugData.

**Comment by porres** [comment]
> Does it have some sort of GUI? 

Nope!

**Comment by porres** [comment]
it'd also be great if my next release would be ready for Pd double precision bits, maybe we can arrange that and also provide PlugData with double precision

**Comment by timothyschoen** [comment]
I'd be down to work on that, I could fix the few things I already noticed, and look for more incorrect usages of "float" instead of "t_float".  Compiler warnings for implicit conversion can help a lot here. 

The consequence of most problems will be very small which makes it very hard to test, especially combined with the fact that ELSE is pretty huge. Telling 32 and 64 bit audio apart by ear is not simple.

**Comment by timothyschoen** [comment]
@porres I didn't see that fluid synth just needs to be tested on M1, I can definitely help with that!

**Comment by porres** [comment]
it's gonna be a little harder than that I guess. Maybe it's easy for you. And maybe you can teach me a few things, can we discuss this on an issue I'm opening on my repository though? See https://github.com/porres/pd-else/issues/1243

**Comment by jaffasplaffa** [comment]
Hey guys :) 

Sorry for my absence, I have had some serious busy weeks. Hopefully during this coming week I will have the time to look a bit more into the documentation again. I have to say I do struggle with some of them, cause I am not 100 % sure how to do it for object with dynamic in and outlet and so on. 

I will read up on it and see if I can make something out of it.

**Comment by timothyschoen** [comment]
No problem, I'll post something here soon about how to continue!

**Comment by kreth608** [comment]
Dropping a comment to say that documentation efforts are under way, and a more up-to date conversation regarding PlugData documentation is available on the [discord channel](https://discord.gg/eT2RxdF9Nq) <3

**Comment by jaffasplaffa** [comment]
Hey everyone!

Sorry I haven't made a lot of effort, besides creating the thread. Too much daytime work and not 100% sure how to do this correctly. I am happy to hear something has happened :) 

Keep up the good work everyone, this is going places, I am sure of :)

**Comment by timothyschoen** [comment]
Documentation discussion now happens in our Discord, and it's getting close to finished as well, so I'm gonna close this. Thanks to anyone who has helped out!
